src/main/java/org/kohsuke/groovy/sandbox/ScopeTrackingClassCodeExpressionTransformer.java

```diff
@@ -108,16 +105,18 @@ def h() {
 
                 Also see issue 17.
              */
-            declareVariable(forLoop.getVariable());
-            super.visitForLoop(forLoop);
+            if (!ForStatement.FOR_LOOP_DUMMY.equals(forLoop.getVariable())) {
+                // When using Java-style for loops, the 3 expressions are a ClosureListExpression and ForStatement.getVariable is a dummy value that we need to ignore.
+                declareVariable(forLoop.getVariable());
+            }
+            forLoop.setCollectionExpression(transform(forLoop.getCollectionExpression()));
+            forLoop.getLoopBlock().visit(this);
```

Comment:
Here we intentionally do not call `super.visitForLoop` because [ClassCodeExpressionTransformer.visitForLoop](https://github.com/apache/groovy/blob/41b990d0a20e442f29247f0e04cbed900f3dcad4/src/main/org/codehaus/groovy/ast/ClassCodeExpressionTransformer.java#L125) transforms the collection expression and then calls `super.visitForLoop(forLoop)`, which will eventually call [`CodeVisitorSupport.visitForLoop`](https://github.com/apache/groovy/blob/41b990d0a20e442f29247f0e04cbed900f3dcad4/src/main/org/codehaus/groovy/ast/CodeVisitorSupport.java#L93) which will visit the collection expression again, but not through `transform`, which we don't want.

Compare that to [`ClassCodeExpressionTransformer.visitIfElse`](https://github.com/apache/groovy/blob/41b990d0a20e442f29247f0e04cbed900f3dcad4/src/main/org/codehaus/groovy/ast/ClassCodeExpressionTransformer.java#L82-L86) which does not itself call `super.visitIfElse`, it instead only visits the constituent parts. I would argue this is a bug in Groovy or at least an inconsistency, but I guess it cannot really be changed without breaking compatibility.